### PR TITLE
o/assertstate: check for conflicts when refreshing and commiting validation set asserts

### PR DESCRIPTION
**Pull request by stolowski**
Refresh validation sets in enforcing mode and check for conflicts before committing them to asserts db.

### CODE REVIEW

File: overlord/assertstate/assertstate.go [hunk 1]
```diff
@@ -356,15 +356,52 @@ func RefreshValidationSetAssertions(s *state.State, userID int) error {
...
+	checkForConflicts := func(db *asserts.Database, bs asserts.Backstore) error {
+		vsets := snapasserts.NewValidationSets()
+		tmpDb := db.WithStackedBackstore(bs)
+		for _, vs := range enforceModeSets {
+			headers := map[string]string{
+				"series":     release.Series,
+				"account-id": vs.AccountID,
+				"name":       vs.Name,
+			}
+			as, err := tmpDb.FindSequence(asserts.ValidationSetType, headers, -1, asserts.ValidationSetType.MaxSupportedFormat())
```
Comment (pedronis) [inline overlord/assertstate/assertstate.go:380]: we never remove assertions, so there might be in principle independently added/left over assertions so I think we need to use the right sequence number and not search for the latest if the validation set is currently pinned

**Review by pedronis — COMMENTED** [review]
looked at this quickly, looks in the right direction, one comment

File: overlord/assertstate/assertstate.go [hunk 2]
```diff
@@ -356,15 +356,60 @@ func RefreshValidationSetAssertions(s *state.State, userID int) error {
...
+				return fmt.Errorf("internal error: unexpected assertion type %s for %s", vsass.Type().Name, ValidationSetKey(vs.AccountID, vs.Name))
+			}
+			if err := vsets.Add(vsass); err != nil {
+				return fmt.Errorf("internal error: cannot check validation sets conflicts: %v", err)
+			}
+		}
+		return vsets.Conflict()
+	}
+
+	enforceModeRefreshErr := bulkRefreshValidationSetAsserts(s, enforceModeSets, checkForConflicts, userID, deviceCtx)
```
Comment (pedronis) [inline overlord/assertstate/assertstate.go:403]: as we discussed if the error is exactly a conflict error we don't want to fail the overall operation, but want to log it and continue (the new validations were not committed though, so we can't update pointers for them below)

File: asserts/pool_test.go [hunk 3]
```diff
@@ -1716,3 +1716,25 @@ func (s *poolSuite) TestPoolReuseWithClearGroupsAndUnchanged(c *C) {
 	})
 	c.Check(toResolveSeq, HasLen, 0)
 }
+
+func (s *poolSuite) TestBackstore(c *C) {
+	assertstest.AddMany(s.db, s.hub.StoreAccountKey(""), s.dev1Acct)
+	pool := asserts.NewPool(s.db, 64)
+
+	c.Assert(pool.AddUnresolved(s.rev1_1111.At(), "for_one"), IsNil)
```
Comment (pedronis) [inline asserts/pool_test.go:1724]: as discussed, if the AtRevision here has Revision: RevisionNotKnown  the test would be slightly clearer. The should be example for that in other tests

Comment (stolowski) [reply asserts/pool_test.go]: Yes. done.

File: overlord/assertstate/assertstate.go [hunk 4]
```diff
@@ -356,30 +356,87 @@ func RefreshValidationSetAssertions(s *state.State, userID int) error {
...
+			if err := vsets.Add(vsass); err != nil {
+				return fmt.Errorf("internal error: cannot check validation sets conflicts: %v", err)
 			}
 		}
+		return vsets.Conflict()
+	}
+
+	if err := bulkRefreshValidationSetAsserts(s, enforceModeSets, checkForConflicts, userID, deviceCtx); err != nil {
+		if _, ok := err.(*snapasserts.ValidationSetsConflictError); ok {
+			logger.Noticef("cannot refresh validation set assertions in enforce mode: %v", err)
```
Comment (pedronis) [inline overlord/assertstate/assertstate.go:433]: maybe "cannot refresh to conflicting validation set assertions ..." ?

**Review by pedronis — APPROVED** [review]
thanks, lgtm, is one of those cases where the diff is hard to follow but the code looks correct afaict. Small remark

**Review by bboozzoo — APPROVED** [review]
LGTM
